File: daemon/src/routes/fs.ts
/**
 * Filesystem routes — GET /fs/complete and GET /fs/check
 *
 * `/fs/complete` backs the directory comboboxes and the Browse dialog with live
 * filesystem suggestions. `/fs/check` reports whether a path exists, is a
 * directory, is a git repo, and has any commits, so the New Agent dialog can
 * describe accurately what registering it will do.
 *
 * Both are read-only, directories-only, capped, and defensive: the daemon binds
 * to 127.0.0.1 and is auth-gated, but these endpoints browse the host
 * filesystem so they stay conservative about what they expose. Neither ever
 * 500s on an unreadable or half-typed path.
 */
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { readdir, stat } from "node:fs/promises";
import { homedir } from "node:os";
import { basename, dirname, isAbsolute, join, sep } from "node:path";
import { hasCommits, isGitRepo } from "../services/git.js";

const MAX_ENTRIES = 50;

const QuerySchema = z.object({
  path: z.string().max(4096),
});

/**
 * Resolve a leading `~` (either exactly `~` or `~/...`) to the user's home
 * dir. `path.join()` normalizes away a trailing separator, which would
 * silently turn `~/` into `/home/user` (no trailing slash) and break the R4
 * "ends with separator → list children" rule for the common home-dir case —
 * so a trailing separator on the input is restored on the output.
 */
function expandTilde(input: string): string {
  if (input === "~") return homedir();
  if (!(input.startsWith(`~${sep}`) || input.startsWith("~/"))) return input;
  const rest = input.slice(2);
  const joined = join(homedir(), rest);
  const hadTrailingSep = input.endsWith(sep) || input.endsWith("/");
  return hadTrailingSep && !joined.endsWith(sep) ? joined + sep : joined;
}

export function registerFsRoutes(app: FastifyInstance): void {
  // GET /fs/check?path=<path>
  app.get("/fs/check", async (req, reply) => {
    const result = QuerySchema.safeParse(req.query);
    if (!result.success) {
      return reply.status(400).send({ error: "Validation error", details: result.error.issues });
    }
    const { path: rawPath } = result.data;

    if (rawPath.includes("\0")) {
      return reply.status(400).send({ error: "Path cannot contain a null byte." });
    }

    const resolved = expandTilde(rawPath);
    if (!isAbsolute(resolved)) {
      return reply.status(400).send({ error: `Path must be absolute (or start with ~). Got: '${rawPath}'` });
    }

    try {
      const stats = await stat(resolved);
      const isDirectory = stats.isDirectory();
      if (!isDirectory) {
        return reply.send({ exists: true, isDirectory: false, isGit: false, hasCommits: null });
      }
      const isGit = await isGitRepo(resolved);
      // Only meaningful for a git dir — project-setup.sh makes an initial
      // commit of the whole directory whenever HEAD doesn't resolve, even for
      // an already-git dir, so callers need this to describe setup accurately.
      const commits = isGit ? await hasCommits(resolved) : null;
      return reply.send({ exists: true, isDirectory: true, isGit, hasCommits: commits });
    } catch {
      // ENOENT/EACCES/etc — user is mid-typing or path doesn't exist. Never 500.
      return reply.send({ exists: false, isDirectory: false, isGit: false, hasCommits: null });
    }
  });

  // GET /fs/complete?path=<partial>
  app.get("/fs/complete", async (req, reply) => {
    const result = QuerySchema.safeParse(req.query);
    if (!result.success) {
      return reply.status(400).send({ error: "Validation error", details: result.error.issues });
    }
    const { path: rawPath } = result.data;

    if (rawPath.includes("\0")) {
      return reply.status(400).send({ error: "Path cannot contain a null byte." });
    }

    const resolved = expandTilde(rawPath);
    if (!isAbsolute(resolved)) {
      return reply.status(400).send({ error: `Path must be absolute (or start with ~). Got: '${rawPath}'` });
    }

    // R4: only list children of the typed dir when the text ends with the
    // path separator. Otherwise prefix-match child names of the parent dir —
    // do NOT auto-descend just because the text already equals a dir name.
    let dir: string;
    let prefix: string;
    if (resolved.endsWith(sep)) {
      dir = resolved;
      prefix = "";
    } else {
      dir = dirname(resolved);
      prefix = basename(resolved);
    }

    let dirents;
    try {
      dirents = await readdir(dir, { withFileTypes: true });
    } catch {
      // ENOENT/EACCES/etc — user is mid-typing, never 500.
      return reply.send({ base: dir, entries: [], truncated: false });
    }

    // Sort by name BEFORE capping, then cap. Sorting the (name-)filtered list
    // is cheap (string compares only, no stat() calls) — it just needs to
    // happen before slice() so the cap keeps the alphabetically-first N
    // entries rather than an arbitrary readdir-order N. Capping before sort
    // (the original approach) is fine for narrow prefix-matched type-ahead
    // (rarely >50 matches) but actively misleading for a full directory
    // listing (empty prefix) in the Browse dialog — e.g. `/usr/lib` could
    // silently omit the very folder the user is looking for.
    // Only directory-ish dirents count toward the cap and toward `truncated`.
    // Filtering on the name alone would let plain files inflate the count: a
    // directory holding 60 files and 2 subdirs would report truncated:true
    // while omitting nothing, and the UI would show a "showing the first 50"
    // warning over a two-item list. Symlinks are kept as candidates because
    // only a stat() can say whether they point at a directory — and that stat
    // stays bounded by the cap below, which is why the cap exists.
    const filtered = dirents
      .filter((d) => prefix === "" || d.name.startsWith(prefix))
      .filter((d) => d.isDirectory() || d.isSymbolicLink())
      .sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
    const truncated = filtered.length > MAX_ENTRIES;
    const candidates = filtered.slice(0, MAX_ENTRIES);

    const entries: { name: string; path: string }[] = [];
    for (const d of candidates) {
      const full = join(dir, d.name);
      if (d.isDirectory()) {
        entries.push({ name: d.name, path: full });
      } else if (d.isSymbolicLink()) {
        try {
          const s = await stat(full);
          if (s.isDirectory()) entries.push({ name: d.name, path: full });
        } catch {
          // Broken symlink — skip.
        }
      }
    }

    return reply.send({ base: dir, entries, truncated });
  });
}

File: rust/vst-routes/src/fs.rs
//! `routes/fs.ts` — GET /fs/complete and GET /fs/check.
//!
//! Ports `daemon/src/routes/fs.ts` (157 LOC):
//! - `GET /fs/check?path=<path>`
//! - `GET /fs/complete?path=<partial>`
//!
//! Read-only, directory-only, capped, and defensive.
//! Expands leading `~` or `~/...` to the user's home directory.

use std::path::{Path, PathBuf};
use vst_agents::home::home_dir;
use vst_git::git::{has_commits, is_git_repo};
use vst_types::rest::fs::{FsCheck, FsComplete, FsCompleteEntry};

pub const MAX_FS_COMPLETE_ENTRIES: usize = 50;
pub const MAX_PATH_LEN: usize = 4096;

/// Errors returned by fs route validation.
#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum FsRouteError {
    #[error("validation_error: path exceeds maximum length of {MAX_PATH_LEN}")]
    PathTooLong,
    #[error("validation_error: path cannot contain a null byte")]
    NullByte,
    #[error("validation_error: path must be absolute (or start with ~). Got: '{0}'")]
    NotAbsolute(String),
}

impl FsRouteError {
    pub fn error_code(&self) -> &'static str {
        "validation_error"
    }
}

/// Resolve a leading `~` (either exactly `~` or `~/...`) to the user's home dir.
/// Preserves trailing slash if present on input.
pub fn expand_tilde(input: &str) -> PathBuf {
    if input == "~" {
        return home_dir();
    }
    if let Some(rest) = input.strip_prefix("~/") {
        let mut base = home_dir();
        if !rest.is_empty() {
            base.push(rest);
        }
        return base;
    }
    PathBuf::from(input)
}

/// Handler for `/fs/*`.
#[derive(Clone, Default)]
pub struct FsRoutes;

impl FsRoutes {
    pub fn new() -> Self {
        Self
    }

    fn validate_path(raw_path: &str) -> Result<PathBuf, FsRouteError> {
        if raw_path.len() > MAX_PATH_LEN {
            return Err(FsRouteError::PathTooLong);
        }
        if raw_path.contains('\0') {
            return Err(FsRouteError::NullByte);
        }

        let resolved = expand_tilde(raw_path);
        if !resolved.is_absolute() {
            return Err(FsRouteError::NotAbsolute(raw_path.to_string()));
        }
        Ok(resolved)
    }

    /// `GET /fs/check?path=<path>`
    pub async fn check(&self, raw_path: &str) -> Result<FsCheck, FsRouteError> {
        let resolved = Self::validate_path(raw_path)?;
        let metadata = match tokio::fs::metadata(&resolved).await {
            Ok(m) => m,
            Err(_) => {
                // ENOENT / EACCES - user is mid-typing or path doesn't exist
                return Ok(FsCheck {
                    exists: false,
                    is_directory: false,
                    is_git: false,
                    has_commits: None,
                });
            }
        };

        if !metadata.is_dir() {
            return Ok(FsCheck {
                exists: true,
                is_directory: false,
                is_git: false,
                has_commits: None,
            });
        }

        let resolved_str = resolved.to_string_lossy();
        let is_git = is_git_repo(&resolved_str).await;
        let commits = if is_git {
            Some(has_commits(&resolved_str).await)
        } else {
            None
        };

        Ok(FsCheck {
            exists: true,
            is_directory: true,
            is_git,
            has_commits: commits,
        })
    }

    /// `GET /fs/complete?path=<partial>`
    pub async fn complete(&self, raw_path: &str) -> Result<FsComplete, FsRouteError> {
        let resolved = Self::validate_path(raw_path)?;

        let has_trailing_sep =
            raw_path.ends_with('/') || raw_path.ends_with(std::path::MAIN_SEPARATOR);

        let (dir, prefix) = if has_trailing_sep {
            (resolved.clone(), String::new())
        } else {
            let parent = resolved
                .parent()
                .map(Path::to_path_buf)
                .unwrap_or_else(|| PathBuf::from("/"));
            let prefix = resolved
                .file_name()
                .map(|f| f.to_string_lossy().to_string())
                .unwrap_or_default();
            (parent, prefix)
        };

        let dir_str = dir.to_string_lossy().to_string();

        let mut read_dir = match tokio::fs::read_dir(&dir).await {
            Ok(rd) => rd,
            Err(_) => {
                // ENOENT/EACCES - user mid-typing, never 500
                return Ok(FsComplete {
                    base: dir_str,
                    entries: vec![],
                    truncated: false,
                });
            }
        };

        let mut candidates = Vec::new();

        while let Ok(Some(entry)) = read_dir.next_entry().await {
            let file_name = entry.file_name().to_string_lossy().to_string();
            if !prefix.is_empty() && !file_name.starts_with(&prefix) {
                continue;
            }

            let file_type = match entry.file_type().await {
                Ok(ft) => ft,
                Err(_) => continue,
            };

            if file_type.is_dir() {
                candidates.push((file_name, entry.path(), true));
            } else if file_type.is_symlink() {
                candidates.push((file_name, entry.path(), false));
            }
        }

        // Sort by name BEFORE capping
        candidates.sort_by(|a, b| a.0.cmp(&b.0));

        let truncated = candidates.len() > MAX_FS_COMPLETE_ENTRIES;
        let selected = if truncated {
            &candidates[..MAX_FS_COMPLETE_ENTRIES]
        } else {
            &candidates[..]
        };

        let mut entries = Vec::new();
        for (name, path, is_known_dir) in selected {
            if *is_known_dir {
                entries.push(FsCompleteEntry {
                    name: name.clone(),
                    path: path.to_string_lossy().to_string(),
                });
            } else {
                // For symlink, check if target is a directory
                if let Ok(meta) = tokio::fs::metadata(path).await {
                    if meta.is_dir() {
                        entries.push(FsCompleteEntry {
                            name: name.clone(),
                            path: path.to_string_lossy().to_string(),
                        });
                    }
                }
            }
        }

        Ok(FsComplete {
            base: dir_str,
            entries,
            truncated,
        })
    }
}
